File: akkaserverpersistent/src/main/scala/com/lightbend/modelserver/actor/queryablestate/QueriesAkkaHttpResource.scala
package com.lightbend.modelserver.actor.queryablestate

import akka.actor.ActorRef
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import com.lightbend.modelServer.model.ModelToServeStats
import com.lightbend.modelserver.actor.actors.GetState
import de.heikoseeberger.akkahttpjackson.JacksonSupport

object QueriesAkkaHttpResource extends JacksonSupport {

  def storeRoutes(modelserver: ActorRef): Route =
    get {
      path("state"/Segment) { datatype =>
        val info: ModelToServeStats = (modelserver ! GetState(datatype)).asInstanceOf[ModelToServeStats]
        complete(info)
      }
    }

}

File: akkaStreamsPersistent/src/main/java/com/lightbend/java/modelserver/actor/queryablestate/QueriesAkkaHTTPResource.java
package com.lightbend.java.modelserver.actor.queryablestate;

import akka.actor.ActorRef;
import akka.http.javadsl.marshallers.jackson.Jackson;
import akka.http.javadsl.server.AllDirectives;
import akka.http.javadsl.server.Route;
import akka.util.Timeout;
import com.lightbend.java.modelserver.actor.actors.GetModels;
import com.lightbend.java.modelserver.actor.actors.GetState;
import com.lightbend.model.ModelServingInfo;

import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

import static akka.http.javadsl.server.PathMatchers.segment;
import static akka.pattern.PatternsCS.ask;

public class QueriesAkkaHTTPResource extends AllDirectives {

    final Timeout askTimeout = Timeout.apply(5, TimeUnit.SECONDS);
    ActorRef router = null;

    public QueriesAkkaHTTPResource(ActorRef router){
        this.router = router;
    }

    public Route createRoute() {
        return route(
                path("models", () ->
                    get(() -> {
                        CompletionStage<List<String>> models = ask(router, new GetModels(), askTimeout).thenApply(result -> (List<String>)result);
                        return completeOKWithFuture(models, Jackson.marshaller());
                    }
                )),
                pathPrefix("state", () ->
                        path(segment(), (String dataType) ->
                                get(() -> {
                                    CompletionStage<ModelServingInfo> info = ask(router, new GetState(dataType), askTimeout).thenApply((ModelServingInfo.class::cast));
                                    return completeOKWithFuture(info, Jackson.marshaller());
                        }))));

    }

}
